mono/utils/mono-state.c

```diff
@@ -927,6 +927,75 @@ mono_native_state_add_memory (MonoStateWriter *writer)
 	mono_state_writer_printf(writer, "},\n");
 }
 
+#define MONO_CRASH_REPORTING_MAPPING_LINE_LIMIT 30
+
+static void
+mono_native_state_add_process_map (MonoStateWriter *writer)
+{
+#if defined(__linux__) && !defined(HOST_ANDROID)
+	static gint32 handle_in_use;
+	gint32 old = mono_atomic_cas_i32 (&handle_in_use, TRUE /* set */, FALSE /* compare */);
+	if (old != FALSE) {
+		g_async_safe_printf ("Warning: Couldn't open /proc/self/maps for concurrent reads, crash report will lack it.\n");
```

Comment:
Why not concurrent? Because of crash reporting, not because of this code?

Reply:
Because it leads to multiple threads trying to open, read from, and close from the same file at the same time. There are enough interesting issues there, that the most straightforward solution (one which doesn't push this interestingness to future debuggers) is just to refuse to do the file IO concurrently.

Reply:
Why can’t the file be opened and read more than once concurrently?

Reply:
That's a good question. It seemed like it was breaking at one point, but that code had multiple bits of unsafe code.

Looking at the POSIX spec, it seems like any Linux kernel after 3.14 should handle that correctly. Let me enable it and see if it works.